commons-numbers-complex/src/main/java/org/apache/commons/numbers/complex/Complex.java

```diff
@@ -20,1290 +20,3678 @@
-        if (isNaN || divisor.isNaN) {
-            return NaN;
+    public static Complex parse(String s) {
+        final int len = s.length();
+        if (len < FORMAT_MIN_LEN) {
+            throw new NumberFormatException(
+                parsingExceptionMsg("Input too short, expected format",
+                                    FORMAT_START + "x" + FORMAT_SEP + "y" + FORMAT_END, s));
         }
 
-        final double c = divisor.getReal();
-        final double d = divisor.getImaginary();
-        if (c == 0.0 && d == 0.0) {
-            return NaN;
+        // Confirm start: '('
+        if (s.charAt(0) != FORMAT_START) {
+            throw new NumberFormatException(
+                parsingExceptionMsg("Expected start delimiter", FORMAT_START, s));
         }
 
-        if (divisor.isInfinite() && !isInfinite()) {
-            return ZERO;
+        // Confirm end: ')'
+        if (s.charAt(len - 1) != FORMAT_END) {
+            throw new NumberFormatException(
+                parsingExceptionMsg("Expected end delimiter", FORMAT_END, s));
         }
 
-        if (Math.abs(c) < Math.abs(d)) {
-            double q = c / d;
-            double denominator = c * q + d;
-            return createComplex((real * q + imaginary) / denominator,
-                (imaginary * q - real) / denominator);
-        } else {
-            double q = d / c;
-            double denominator = d * q + c;
-            return createComplex((imaginary * q + real) / denominator,
-                (imaginary - real * q) / denominator);
+        // Confirm separator ',' is between at least 2 characters from
+        // either end: "(x,x)"
+        // Count back from the end ignoring the last 2 characters.
+        final int sep = s.lastIndexOf(FORMAT_SEP, len - 3);
+        if (sep < BEFORE_SEP) {
+            throw new NumberFormatException(
+                parsingExceptionMsg("Expected separator between two numbers", FORMAT_SEP, s));
         }
-    }
 
-    /**
-     * Returns a {@code Complex} whose value is {@code (this / divisor)},
-     * with {@code divisor} interpreted as a real number.
-     *
-     * @param  divisor Value by which this {@code Complex} is to be divided.
-     * @return {@code this / divisor}.
-     * @see #divide(Complex)
-     */
-    public Complex divide(double divisor) {
-        if (isNaN || Double.isNaN(divisor)) {
-            return NaN;
+        // Should be no more separators
+        if (s.indexOf(FORMAT_SEP, sep + 1) != -1) {
+            throw new NumberFormatException(
+                parsingExceptionMsg("Incorrect number of parts, expected only 2 using separator",
+                                    FORMAT_SEP, s));
         }
-        if (divisor == 0d) {
-            return NaN;
+
+        // Try to parse the parts
+
+        final String rePart = s.substring(1, sep);
+        final double re;
+        try {
+            re = Double.parseDouble(rePart);
+        } catch (final NumberFormatException ex) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)
Similar issue at line numbers 417.